backend/src/external_api/cdogs/cdogs.service.ts

```diff
@@ -0,0 +1,192 @@
+    //-- that the the template hash is still valid
+    if (config?.configurationValue) {
+      result = await this._isCachedTemplateValid(apiToken, config.configurationValue);
+    }
+
+    return result;
+  };
+
+  //--
+  //-- upload a template to the CDOGS api, if the template exists
+  //-- the system will return the template hash, and this will
+  //-- need to be stored for the next time the document needs to be
+  //-- generated
+  //--
+  upload = async (apiToken: string, type: string, templateCode: string) => {
+    const url = `${this.baseUri}/api/v2/template`;
+
+    const template =
+      type === "HWCR"
+        ? "templates/complaint/CDOGS-HWCR-COMPLAINT-TEMPLATE-v1.docx"
+        : "templates/complaint/CDOGS-ERS-COMPLAINT-TEMPLATE-v1.docx";
+    const path = join(process.cwd(), template);
+
+    try {
+      const bodyFormData = new FormData();
+      bodyFormData.append("template", fs.createReadStream(path));
+      const headers: any = {
+        ...bodyFormData.getHeaders(),
+      };
+
+      const response: AxiosResponse = await post(apiToken, url, bodyFormData, headers);
+
+      //-- update the system with the hash provided by the CDGOS api
+      const { data: hash } = response;
+      this.configService.updateByCode(templateCode, hash);
+    } catch (error) {
+      //-- if CDOGS already has a template cached the api will throw an error
+      //-- with the existing template hash, this will need to be saved for
+      //-- the next time the document is generated
+      const {
+        response: {
+          data: { status, hash },
+        },
+      } = error;
+
+      if (status === constants.HTTP_STATUS_METHOD_NOT_ALLOWED) {
+        //-- existing template
+        this.configService.updateByCode(templateCode, hash);
+      } else {
+        //-- valid error
+      }
+    }
+  };
+
+  //--
+  //-- render complaint to pdf
+  //--
+  generate = async (documentName: string, data: any, type: COMPLAINT_TYPE): Promise<AxiosResponse> => {
+    const templateCode = type === "HWCR" ? CONFIGURATION_CODES.HWCTMPLATE : CONFIGURATION_CODES.ERSTMPLATE;
+
+    try {
+      const apiToken = await this.authenticate();
+
+      if (!(await this.isTemplateCached(apiToken, templateCode))) {
+        await this.upload(apiToken, type, templateCode);
+      }
+
+      const uid = await this._getTemplateId(templateCode);
+      const url = `${this.baseUri}/api/v2/template/${uid}/render`;
+
+      const config: AxiosRequestConfig = {
+        responseType: "arraybuffer",
+        headers: {
+          "Content-Type": "application/json",
+          Authorization: `Bearer ${apiToken}`,
+        },
+      };
+
+      const documentData = await this._applyData(data, documentName);
+
+      const response = await axios.post(url, documentData, config);
+      return response;
+    } catch (error) {
+      console.log(`exception: unable to export document for complaint: ${data.id} - error: ${error}`);
```

Comment:
We should be using the Logger middleware.

e.g. private readonly logger = new Logger(CdogsService.name);

then this.logger.error(bla)

There are a few reasons, but in this case we have the same exception text in several classes, and the console log statement doesn't tell us which class throws this error in the logs.

Reply:
Thanks for catching that, I totally forgot about the logger, and you're absolutely right